### PR TITLE
Fix error raised on invalid JSON

### PR DESCRIPTION
**Pull request by clokep**
One of my previous PRs (#8106) had a bug in it -- you can't easily raise a `JSONDecodeError`, so this just changes it to a `ValueError`.

While I was in there I wrote some (very simple) tests for the helpers used to deserialize JSON since I didn't see any!

I targeted the `release-v1.20.0` branch since this change was not yet released.

### CODE REVIEW

**Review by richvdh — APPROVED** [review]
lgtm

**Comment by clokep** [comment]
Tests are failing due to #8285 not being on the release branch. I've pushed a branch to sytest which is equivalent to sytest's `release-v.1.20.0`.

File: synapse/http/matrixfederationclient.py [hunk 1]
```diff
@@ -164,7 +165,8 @@ async def _handle_json_response(
     try:
         check_content_type_is_json(response.headers)
 
-        d = treq.json_content(response)
+        # Reject Python extensions to JSON.
+        d = treq.json_content(response, parse_constant=_reject_invalid_json)
```
Comment (clokep) [inline synapse/http/matrixfederationclient.py:169]: Looks like this requires upping the minimum version of treq to 18.6.0. Do we have any issues with that?

Comment (clokep) [reply synapse/http/matrixfederationclient.py]: Fedora 31 (EOL Nov. 17) only has treq 17.8.0, so I decided to essentially [inline part of `json_content`](https://github.com/twisted/treq/blob/dcd63574b574595b30beb7b442329e995181f6a3/src/treq/content.py#L99-L101).

**Comment by clokep** [comment]
@richvdh Could you take another look since I made some fixes for old-deps? Thanks!

**Review by richvdh — APPROVED** [review]
lgtm

**Pull request merged**